### PR TITLE
Add the `point-back` feature

### PR DESCRIPTION
**Pull request by lohsuan**
## Description
As title

## How to verify
1. import token in meta mask
- `0xFD63AC01eb9D85a5fbdB5E669BE4D7CA7136f29d`
- 
- image
2. donate

### CODE REVIEW

**Review by Xanonymous-GitHub — COMMENTED** [review]
Suggest changing all Openzepplin-defined IERC20 methods to be an `override` method instead of `virtual`.

**Comment by tzulin** [comment]
Okay. Just a minute.

**Review by Xanonymous-GitHub — COMMENTED** [review]
And also suggest removing all `virtual` flag from our Token contract methods since we may not need to allow others to override our contract methods.

Then, suggest strictly using `external`, `internal` instead of `public` for our contract methods (of course `private` is the best).

**Comment by tzulin** [comment]
Okay

File: src/assets/constants/puppySponsorContract.json [hunk 1]
```diff
@@ -1,6 +1,17 @@
 {
-  "CA": "0x711672C59393883240276df8d68de897C11e5B94",
+  "CA": "0x7fdFE5Baf61aEe72A1B60Af76f6a50d6d4953098",
```
Comment (Xanonymous-GitHub) [inline src/assets/constants/puppySponsorContract.json]: [Question]
Is this our current contract address?

File: contracts/PuppyToken.sol [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+     * - `account` cannot be the zero address.
+     * - `account` and `totalSupply` must have at least `amount` tokens.
+     */
+    function burn(address account, uint256 amount) external onlyOwner {
+        require(account != address(0), "PuppyToken: burn from the zero address");
+        require(totalSupply >= amount, "PuppyToken: burn amount exceeds the total supply");
+        require(_balances[account] >= amount, "PuppyToken: burn amount exceeds the balance of account");
+
+        require(totalSupply - amount <  totalSupply, "PuppyToken: burn overflow");
+        require(_balances[account] - amount <  _balances[account], "PuppyToken: burn overflow");
```
Comment (Xanonymous-GitHub) [inline contracts/PuppyToken.sol]:
```suggestion
        uint256 accountBalance = _balances[account];
        require(accountBalance >= amount, "PuppyToken: burn amount exceeds the balance of account");

        require(totalSupply - amount <  totalSupply, "PuppyToken: burn overflow");
        require(accountBalance - amount <  accountBalance, "PuppyToken: burn overflow");
```

Suggest not to repeatedly access elements in the array, if possible, store the value in advance to reduce unnecessary gas fees.

File: contracts/PuppyToken.sol [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        address from,
+        address to,
+        uint256 amount
+    ) private {
+        require(from != address(0), "PuppyToken: transfer from the zero address");
+        require(to != address(0), "PuppyToken: transfer to the zero address");
+        require(_balances[from] >= amount, "PuppyToken: transfer amount exceeds balance");
+        
+        require(_balances[to] + amount > _balances[to], "PuppyToken: transfer overflow");
+        require(_balances[from] - amount < _balances[from], "PuppyToken: transfer overflow");
```
Comment (Xanonymous-GitHub) [inline contracts/PuppyToken.sol]: ditto.

**Review by Xanonymous-GitHub — COMMENTED** [review]
Seems our Token contract does not fit to ERC20 standard yet, we should add the missing `transferFrom` method.
```solidity
function transferFrom(address from, address to, uint256 amount) public returns (bool)
```

**Comment by Xanonymous-GitHub** [comment]
> **Warning**
Later, I checked the ERC20 standard and found that many methods are required to be declared as `public` instead of `external`, so I think it is necessary for us to check again whether the contract interface fully conforms to the ERC20 standard.

https://ethereum.org/en/developers/docs/standards/tokens/erc-20/
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/IERC20.sol

I haven't thought deeply about this a while ago, and I'm very, very sorry for the useless advice (https://github.com/lohsuan/puppy_sponsor/pull/8/files#pullrequestreview-1008733848).... 😢😢😢😢😢😢😢😢😢😢😢😢😢😢😢😢😢😢😢😢 

So that means we should check the standard again, and(maybe) change some methods from `external` to `public`.

File: src/assets/constants/puppySponsorContract.json (hunk 1 above)
Comment (lohsuan) [reply src/assets/constants/puppySponsorContract.json]: It is the version before the earlier modified.
I think the newest contract is not deployed yet.

Comment (Xanonymous-GitHub) [reply src/assets/constants/puppySponsorContract.json]: Ok, then kindly remind to replace this value after the final deployment :)

Comment (lohsuan) [reply src/assets/constants/puppySponsorContract.json]: > **Note**
> Redeploy Contract :)

Comment (Xanonymous-GitHub) [reply src/assets/constants/puppySponsorContract.json]: > **Note**
> LGTM

File: contracts/PuppyToken.sol (hunk 2 above)
Comment (lohsuan) [reply contracts/PuppyToken.sol]: resolve here [1f18804](https://github.com/lohsuan/puppy_sponsor/pull/8/commits/1f18804523a38a4bc276859b1f5b9e0c96fc8b0b)

File: contracts/PuppyToken.sol (hunk 3 above)
Comment (lohsuan) [reply contracts/PuppyToken.sol]: resolved here [1f18804](https://github.com/lohsuan/puppy_sponsor/pull/8/commits/1f18804523a38a4bc276859b1f5b9e0c96fc8b0b)

**Comment by Xanonymous-GitHub** [comment]
This PR has a branch base problem need to be fixed, please don't merge before rebase.

**Review by Xanonymous-GitHub — APPROVED** [review]
LGTM.

> **Note**
The branch wrong-base issue has been fixed, and also kindly remind changing the CA value after re-deploy contracts.

**Review by tzulin — APPROVED** [review]
LGTM. Everything worked well.
Thank you!

Oh, wait. I saw the note. Give me a second to re-deploy it.

**Pull request merged**